Adds L2 Bridge resources for Arbitrum

wackerow opened the pull request:
### `AboutLayer2.vue`
- Fills in the section on getting Arbitrum funds on the `AboutLayer2` page. 
- Adds logic to check if Arbitrum is being used for current round, hiding content if not
- Uses string interpolation for native token references
- Button placed at top of `AboutLayer2` page to quick-scroll down to bridging information
- Currently, if NOT on Arbitrum, L2 bridging information will be hidden on this page, as the copy here is specific to Arbitrum.

### `Profile.vue`
- Adds a warning to the `Profile` component if either balance is zero (ETH or ERC-20)
- Adjusts the warning if _only one_ balance is zero, to emphasize/clarify to the user that both are needed.
- Warning provides link to `AboutLayer2` page

### `chains.ts`, `api/core.ts`
- Added a `bridge` property to `CHAIN_INFO` within `chains.ts` to allow reusable chain-specific link
- Updated `api/core.ts` to utilize explorer data in `chains.ts`, and removed `VUE_APP_BLOCK_EXPLORER` from `vue-app/.env` (removes redundancy)
- Adds a `isLayer2` property to easily determine if we should show layer 2 resources or not elsewhere in the app
- Added Optimism, Polygon and xDai chain details with assets for potential future use
- Added `currency` property to account for chains that don't necessarily use ETH (ie Polygon (MATIC) and xDai (xDai))... This would still need to be rolled out to rest of app as part of the upstream merge project. 

### `Landing.vue`
- Updates "Arbitrum" CTA to be dynamic based on chain (sourced by chain ID in `vue-app/.env`)
- If chain is a layer 2, link will direct to `AboutLayer2` resources which connects them to the bridge
- Else, it will connect directly to the bridge declared in `chains.ts`

### Related issue
#428 (Hasn't fixed entirely yet, can address remaining instances of hard-coded "Arbitrum" in separate PR)
Fixes: [Notion card](https://www.notion.so/efdn/Add-L2-bridge-resources-78d1df28acbc4ea49c73d0088fa81f90)

wackerow committed:
add `bridge` property to ChainInfo

wackerow committed:
adds Arbitrum bridge info to AboutLayer2

wackerow committed:
adds warning to Profile if no ETH or ERC-20

links to AboutLayer2 page for bridge info

wackerow committed:
pull blockExplorer from `chains.ts` depending on chainId

Removes need for `VUE_APP_BLOCK_EXPLORER` in `.env`, but can optionally override explorer link by using one.

wackerow committed:
eslint ignore for error catch

wackerow committed:
remove redundant `chainLabel` getter

wackerow committed:
fix blockExplorer constant, removes from .env

(`VUE_APP_BLOCK_EXPLORER` removed from `.env`)

wackerow committed:
updates `chains.ts` list

Adds Optimism, xDai and Polygon details to list. Adds `currency` parameter for user with networks such as xDai and Polygon (which do not use ETH). Adds `isLayer2` parameter to help identify when Layer 2 resources should be displayed for user.

wackerow committed:
makes landing L2 content dynamic based on chain

samajammin commented on vue-app/src/api/core.ts:
```diff
@@ -9,7 +10,7 @@ export const provider = new ethers.providers.StaticJsonRpcProvider(
 )
 export const blockExplorer =
-  process.env.VUE_APP_BLOCK_EXPLORER || 'https://etherscan.io'
+  CHAIN_INFO[Number(process.env.VUE_APP_ETHEREUM_API_CHAINID)].explorer
```
Might be nice to have a check here to make sure a valid `VUE_APP_ETHEREUM_API_CHAINID` is used, similar to how we check against registry types, e.g. https://github.com/ethereum/clrfund/blob/develop/vue-app/src/api/core.ts#L35

samajammin commented on vue-app/src/views/AboutLayer2.vue:
```diff
@@ -57,33 +64,159 @@
     </p>
+    <!-- If chain is Arbitrum, display bridge information: -->
+    <div
+      v-if="chain.label.includes('Arbitrum')"
```
If we find ourselves having to do this in multiple places, might be useful to just make a util function e.g. in networks.ts for this.

samajammin commented on vue-app/src/views/AboutLayer2.vue:
```diff
@@ -57,33 +64,159 @@
+            and select token.
+          </p>
+        </li>
+        <li>Enter amount & click "Deposit"</li>
```
```suggestion
        <li>Enter amount and click "Deposit"</li>
```

samajammin commented on vue-app/src/views/AboutLayer2.vue:
```diff
@@ -57,33 +64,159 @@
+    return (window as any).ethereum
+  }
+
+  async addTokenToWallet() {
```
Cool!

samajammin commented on vue-app/src/views/AboutLayer2.vue:
```diff
@@ -57,33 +64,159 @@
+
+  async addTokenToWallet() {
+    try {
+      if (this.windowEthereum) {
```
I'm guessing this functionality only works for MetaMask? Should we add a check e.g. `this.windowEthereum && this.windowEthereum.isMetaMask`?

wackerow replied:
Perhaps, not sure honestly... does `isMetaMask` stay true if using WalletConnet while linked to MetaMask?

wackerow replied:
Just tested it out, and the button did nothing when connected to MetaMask through WalletConnect on mobile. Will add the conditional for `isMetaMask`

samajammin commented on vue-app/src/views/Landing.vue:
```diff
@@ -204,6 +210,16 @@ export default class Landing extends Vue {
       ?.scrollIntoView({ behavior: 'smooth' })
   }
+
+  get chain(): any {
```
Same deal here - if we need this in multiple places, should we just create util functions in networks.ts?

wackerow replied:
Hey @samajammin, take a peak at the last two commits (4e42c01 and 1142882) and let me know your thoughts... 

I tried to consolidate the recurrent `CHAIN_INFO[Number(process.env.VUE_APP_ETHEREUM_API_CHAINID)]` lookups by just fetching this in `api/core`, checking if it's a valid chainId, and exporting the chain as `chain`.

We can then use the data properties on the `chain` object for easy reference in the front end after a simple import and getter, ie:
```js
import { chain } from '@/api/core'
// ...
get chain {
  return chain // this makes it accessible within the Vue template
}
```

Is this kinda what you had in mind? This seems to be one of the simpler ways to organize all this chain data so it's easily accessible. Feel free to suggest otherwise.

wackerow replied:
Also, side note related to `networks.ts`... noticing it really only checks for XDAI and we now have a setup in `chains.ts` that can return the chains native currency. I think we should patch this so it looks up from `chains.ts` and returns `.currency`... trying to draw the line on this PR though.

pettinarip replied:
> Also, side note related to `networks.ts`... noticing it really only checks for XDAI and we now have a setup in `chains.ts` that can return the chains native currency. I think we should patch this so it looks up from `chains.ts` and returns `.currency`... trying to draw the line on this PR though.

Yes, that makes sense.

samajammin commented on vue-app/src/views/Profile.vue:
```diff
@@ -129,7 +143,14 @@ import { CHAIN_INFO, ChainInfo } from '@/plugins/Web3/constants/chains'
+    Links,
+  },
 })
 export default class NavBar extends Vue {
```
```suggestion
export default class Profile extends Vue {
```
Lol

samajammin commented:
@wackerow one additional place I noticed is the NavBar component - please add logic there to hide the "Layer 2" dropdown link if the app isn't running on L2.

samajammin reviewed (changes requested):
Overall this is looking awesome! Nice work.

Left a few comments.

wackerow committed:
Apply suggestions from code review

Co-authored-by: Sam Richards <sbrichards@gmail.com>

wackerow committed:
simplify `chain` export

Add check for valid `chain` and throw error if not. Remove `blockExplorer` export, simply using `chain.explorer`

wackerow committed:
implements `chain` import usage in .vue files

CHAIN_INFO lookup is exported from core api and holds data properties that can be accessed easily in front-end. Makes naming consistent across components with just `chain`

wackerow committed:
adds fallback bridge info if `bridge` url available

If not on Arbitrum

wackerow committed:
make /layer-2 link conditional in dropdown

wackerow committed:
enable param to auto-scroll to bridge section

Can direct to Layer2 page with param to automatically scroll to the bridge section for more directed guidance.

wackerow committed:
added L2 conditional to copy of About

wackerow committed:
adds L2 conditionals to recipient guide

checks if `chain.bridge` is available before rendering guide to get funds. Adds check for `isLayer2` before showing link to Layer2 page. Instead of direct link to Arbitrum tutorial, made more dynamic by linking to the `bridge` section of the Layer2 page

wackerow committed:
Create FundsNeededWarning.vue

Extracted from Profile for re-use. Removed $warning-color border and adjusted margin on warning div.

wackerow committed:
implement FundsNeededWarning into Cart and Profile

wackerow committed:
clean up and add use of `ChainInfo` interface

wackerow committed:
adds `isCompact` prop

Allows optionally showing a limited form of the warning, for display under the contribute button in cart.

wackerow committed:
adds isMetaMask check for addTokenToWallet

wackerow committed:
block explorer url patch

wackerow committed:
mock arbitrum for hardhat chain properties

wackerow committed:
add links for arbitrum/walletconnect supported wallets

wackerow committed:
adds conditionals for L2 links

pettinarip commented on vue-app/src/components/FundsNeededWarning.vue:
```diff
@@ -0,0 +1,110 @@
+<template>
+  <div :class="!isCompact && 'warning'" v-if="needsFunds">
+    <span v-if="!isCompact">You need L2 funds!</span>
```
Shouldn't this text be behind the `chain.isLayer2` flag?

wackerow replied:
Yep, agree.. patching =)

pettinarip commented on vue-app/src/components/FundsNeededWarning.vue:
```diff
@@ -0,0 +1,110 @@
+        Get help bridging {{ singleTokenNeeded }} to Layer 2
+      </links>
+      <links v-else :to="chain.bridge">
+        Bridge {{ singleTokenNeeded }} to Layer 2
```
Not sure if I understand this logic. If we are in this condition is because `chain.isLayer2` is `false`, so the text should be different, right? without mentioning "Layer 2".

wackerow replied:
Good catch! Patching

pettinarip commented on vue-app/src/views/AboutLayer2.vue:
```diff
@@ -57,33 +64,203 @@
+        class="btn-secondary"
+        @click="addTokenToWallet"
+      >
+        Add {{ chain.label }} {{ nativeToken.symbol }} to MetaMask
```
Oh yeah! would be nice to have an open source package (as a plugin) for these buttons :)

pettinarip commented on vue-app/src/views/Profile.vue:
```diff
@@ -120,18 +121,28 @@ import IconStatus from '@/components/IconStatus.vue'
+  },
 })
-export default class NavBar extends Vue {
+export default class Profile extends Vue {
```
Haha thinking for how long is it that we had it like that

pettinarip commented on vue-app/src/views/Profile.vue:
```diff
@@ -369,4 +384,16 @@ p.no-margin {
+  @media (max-width: $breakpoint-s) {
+    margin: 1rem 0 0;
+  }
+} */
```
cleanup?

pettinarip reviewed (commented):
Really good job here @wackerow. Added a couple of comments.

wackerow committed:
fix l2 copy logic

wackerow committed:
removed code cleanup

samajammin reviewed (approved):
🏆

samajammin committed:
Merge branch 'develop' into l2-resources
